### ISSUE TITLE
Sanitize() function strips out necessary escaped double quotes

### ISSUE THREAD

**Issue by leifmadsen** [redhat-service-assurance/smart-gateway#36]
The `Sanitize()` function in `pkg/internal/saelastic/utils.go` results in the stripping of escaped double quotes when that is not what we want to happen in certain cases.

https://github.com/redhat-service-assurance/smart-gateway/blob/1f62330068b08fafa937b8285bc4cb35f1e10205/internal/pkg/saelastic/utils.go#L17-L27

This block of code seems to swap out `\"` with `"` which is probably fine in multiple cases, but when you have a block of JSON like so:

```
[{"labels":{"alertname":"collectd_interface_if_octets","instance":"localhost.localdomain","interface":"lo","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"Host localhost.localdomain, plugin interface (instance lo) type if_octets: Data source \"rx\" is currently 43596.224329. That is above the failure threshold of 0.000000.","DataSource":"rx","CurrentValue":"43596.2243286703","WarningMin":"nan","WarningMax":"nan","FailureMin":"nan","FailureMax":"0"},"startsAt":"2019-09-18T21:11:19.281603240Z"}]
```

And then process it with `Sanitize()` we end up with the following:

```
{"labels":{"alertname":"collectd_interface_if_octets","instance":"localhost.localdomain","interface":"lo","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"Host localhost.localdomain, plugin interface (instance lo) type if_octets: Data source "rx" is currently 43596.224329. That is above the failure threshold of 0.000000.","DataSource":"rx","CurrentValue":"43596.2243286703","WarningMin":"nan","WarningMax":"nan","FailureMin":"nan","FailureMax":"0"},"startsAt":"2019-09-18T21:11:19.281603240Z"}
```
Then you end up with an issue with the JSON. If we take this event after processing it through `Sanitize` and try to process it with a JSON tool in python to pretty print it, then we end up with the following. Looking at column 264, we are looking at the `summary` field with the unescaped `"rx"`.

```
echo '{"labels":{"alertname":"collectd_interface_if_octets","instance":"localhost.localdomain","interface":"lo","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"Host localhost.localdomain, plugin interface (instance lo) type if_octets: Data source "rx" is currently 43596.224329. That is above the failure threshold of 0.000000.","DataSource":"rx","CurrentValue":"43596.2243286703","WarningMin":"nan","WarningMax":"nan","FailureMin":"nan","FailureMax":"0"},"startsAt":"2019-09-18T21:11:19.281603240Z"}' | python -m json.tool
Expecting , delimiter: line 1 column 264 (char 263)
```

If we process the raw initial event, then we get our desired output, where we can see the `summary` field containing the desired escaped `\"rx\"` which can be processed without issue.

```
echo '[{"labels":{"alertname":"collectd_interface_if_octets","instance":"localhost.localdomain","interface":"lo","severity":"FAILURE","service":"collectd"},"annotations":{"summary":"Host localhost.localdomain, plugin interface (instance lo) type if_octets: Data source \"rx\" is currently 43596.224329. That is above the failure threshold of 0.000000.","DataSource":"rx","CurrentValue":"43596.2243286703","WarningMin":"nan","WarningMax":"nan","FailureMin":"nan","FailureMax":"0"},"startsAt":"2019-09-18T21:11:19.281603240Z"}]' | python -m json.tool
[
    {
        "annotations": {
            "CurrentValue": "43596.2243286703",
            "DataSource": "rx",
            "FailureMax": "0",
            "FailureMin": "nan",
            "WarningMax": "nan",
            "WarningMin": "nan",
            "summary": "Host localhost.localdomain, plugin interface (instance lo) type if_octets: Data source \"rx\" is currently 43596.224329. That is above the failure threshold of 0.000000."
        },
        "labels": {
            "alertname": "collectd_interface_if_octets",
            "instance": "localhost.localdomain",
            "interface": "lo",
            "service": "collectd",
            "severity": "FAILURE"
        },
        "startsAt": "2019-09-18T21:11:19.281603240Z"
    }
]
```

This issue will require that the `Sanitize()` function be cleaned up and made more robust to account for these types of situations when processing events.

**Comment by leifmadsen** [comment]
@paramite, @aneeshkp any suggestions on a better way forward here? I wonder if there is a better way to process and sanitize the inbound events while allowing for guaranteed valid JSON output.

**Comment by leifmadsen** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Sanitize() function results in invalid JSON within the Smart Gateway saelastic package](https://trello.com/c/RXTFuLSr/669-sanitize-function-results-in-invalid-json-within-the-smart-gateway-saelastic-package)

**Comment by aneeshkp** [comment]
@leifmadsen  yea this is used for events . In collectd  VES is sent as a string and it was breaking elastic search . I believe @abays  has a fix for it in collectd. But you are right meanwhile we can try to make this more robust in smart gateway .

**Comment by leifmadsen** [comment]
I found this (https://github.com/OWASP/json-sanitizer) last night, which made me wonder if having something like that could be useful. Might make the inbound and outbound JSON messages a bit cleaner and sane.

Not saying this is the answer, but the idea of something that can clean up inbound JSON and sanitize it on the outbound to ElasticSearch seems like a sane approach.

**Comment by leifmadsen** [comment]
This issue is related to https://github.com/collectd/collectd/issues/2697 where we're trying to sanitize some of the inbound events but cause a regression.

**Comment by leifmadsen** [comment]
Closed by #38
